### ISSUE TITLE
RAR FILES SUPPORT 

### ISSUE THREAD

**Issue by NerdFlix** [zach-morris/plugin.program.iarl#39]
Hi Zach, first of all thanks for your awesome addon.
I have a custom games list and majority of them are in .rar format, can you please provide rar support as an enhancement ?? A post download option that detects the file extension(rar,zip,etc) and extract the file would be welcome as well for those who have a custom list with several different files format.

Thanks in advance.

**Comment by zach-morris** [comment]
That should be pretty straightforward.  I believe Kodi already supports RAR files.

**Comment by NerdFlix** [comment]
Really? good to hear that, I'm looking forward for this enhancement. I have custom lists for Gamecube, PS2, Nintendo DS and PSP. The download links are from Emuparadise, there is a scraper to get the links, I've made some changes in IARL to be able to download them, the thing is that you won't know the file extensions until you request the files to download. Let me know if you are interested to bring Emuparadise games to IARL as it is very simple.

Thanks zach :)

**Comment by zach-morris** [comment]
RAR support shouldn't be a problem.  I'm going to leave custom lists like emuparadise to the user however to avoid any issues with DMCA.  I'll try to get something together in the next week or so with RAR support, which would mimic the zipfile support functions already present.

**Comment by NerdFlix** [comment]
Ok Zach, thanks

**Comment by NerdFlix** [comment]
Hi zach,
Does Kodi support RAR5 files? I think their built in function doesn't unrar Rar5 files

**Comment by zach-morris** [comment]
I'm not sure.  I haven't tried it yet.  If not, then we might be stuck and have the same issue as 7zip.  7zip support I can get to work for all platforms except android, since android does not allow binary packages to be included in the addon (currently).  7zip apparently supports rar5 and is already included in IARL, and I believe that works for all platforms except android.

**Comment by NerdFlix** [comment]
Luckily for me I'm using Windows and Kodi built in function can unrar all my rar files( I suppose they're not Rar5 files). I'm stuck because my custom list is not single file format, there are many roms in rar,  zip, 7z, etc, and I don't know if a funtion to detect the file format and decompress using the right tool is a need for all users( or just mine), unlike rar support. Anyway, I appreciate all enhancements you bring to us.

**Comment by zach-morris** [comment]
Currently IARL is not setup to handle non-standard lists of file types.  They pretty much all have to be the same.  See [here](https://github.com/zach-morris/plugin.program.iarl/wiki/3.--Create-your-own-lists).  After a file is downloaded, the addon will perform the post download action defined to process the file however it's needed, but it will do that for every file in the xml list.  For RAR files, my plan was just to add a post download action for unrar and then point to the one file thats extracted.  This is the same thing the addon does now for zipfiles.  If the files weren't standardized, then there wouldn't be an easy way to set this up.

**Comment by NerdFlix** [comment]
Yeah you're right, as I said a function to do that is just my need, I copied your code and made a function that verifies the extensions and decompress the roms using the right tool, I created a new post download option to not affect the standardized lists. I didn't test it enough, but for now it is decompressing the roms.

**Comment by zach-morris** [comment]
The more I look at this, the more I think it's currently not going to be possible to create a platform independent solution for rar/rar5 files.  However, here's a thought:
If I create a new arbitrary "manual command" postdl action (similar to the "manual command" external launch command).  That way, IARL will simply execute whatever arbitrary command you're looking for.

Example in the game xml would include:
`<emu_postdlaction>execute_manual_postdlaction('"C:\Program Files\WinRAR\WinRAR.exe" x %DOWNLOADED_FILE% %EMU_DOWNLOADPATH%')</emu_postdlaction>`

This new advanced command would need to support a few wildcards:
DOWNLOADED_FILE:  Obviously the file that was downloaded that needs to be processed
EMU_DOWNLOADPATH:  The current archives download path setting, so the external command would know where to extract the file to

I can't think of any other wildcards at the moment that would be needed... If you think of some let me know.

Then the only other caveat is that whatever file is extracted would be assumed to be named the same.  Ex:  Mario.rar extracts to Mario.nes
Then IARL could search for the same filename and try and launch accordingly.

Thoughts?

**Comment by zach-morris** [comment]
rarfile has the same issue:

> Compressed files are extracted by executing external tool: either unrar from RARLAB or bsdtar from libarchive.

patool is a similar python archive library which supports everything under the sun, which is what originally I was interested in including in the addon when 7zip became an issue, but again it has the same drawbacks.  In general, as far as I can tell, the only archive scheme that does not rely on an external binary is zipfile which is supported natively in python.

Ultimately, the solution for archive file formats will be handled by kodi directly.  I believe in Kodi v18, a [VFS](https://github.com/notspiff/vfs.rar) will be implemented to include support for all these various formats.

**Comment by NerdFlix** [comment]
If Kodi will handle all these file format mess, it will make everything simpler, their native function today for big  non rar5 files(gamecube and PS2 ISOs) cannot be trusted and is very slow.

I liked the manual command solution, it gives a customization option not even for rar files, but post download option overall. One step ahead would be to give the option to the user to run his own function as long as it respects some wildcards and return values, as the same rom name with only different extension.

You're not even close to be a clueless dev, actually you did a pretty good job with this addon and IARL is by far the best option for rom management for different OS.

I'm not even a dev so my thoughts should be taken as simple end user opinions. Any solution you give for rar support will be welcome and thanks for your effort.

**Comment by zach-morris** [comment]
If possible, send me a PM in the Kodi forums.  I need a few test cases to check

**Comment by zach-morris** [comment]
I'm not sure whatever solution I come up with is going to completely solve your issue.  Nevertheless I think adding a more generalized manual post download action is a good thing to do.

The main issue I see is the files are somewhat non-deterministic.   We won't know ahead of time what the file we should attempt to launch will be named.  The assumption as suggested before is that the file is named exactly the same as the downloaded file:
e.g. mario.rar will extract to mario.rom

But as I see it, mario.rar might extract to mario.rar1... mario.rar99 which then have to be further extracted from parts into mario.rom

Likely the solution will have to be an interim script that does some massaging of the file.  My suggestion would be:
1)  Generalized post DL action is called:
`<emu_postdlaction>execute_manual_postdlaction('"C:\Program Files\...\recursive_decompress.bat" %DOWNLOADED_FILE%"')</emu_postdlaction>`

2)  The DL action calls a bat file, which would uncompress and rename the file appropriately, and then delete the RAR.  See [here](https://stackoverflow.com/questions/8976287/recursive-unzipping-with-7z-exe
) for a possible solution.

3)  IARL then continues on and attempts to find and launch the %DOWNLOADED_FILE% with a different extension (using the same matching technique IARL already uses to see if the game file is already present)

The other issue I see is this decompression is a completely different thread for the computer, so there's potentially no feedback the decompression process is completed before IARL continues and then attempts to launch the game.  I'm not really sure how to overcome that hurdle.

**Comment by NerdFlix** [comment]
Yes you're right, the main issue with custom lists is that we won't know ahead of time what file to launch after uncompression. 

Did you test the code I sent to you? Is it working for my list on your tests? by working I mean, extracting and merging splitted files if needed and renaming to the same as the downloaded file with new extension, and deleting the rar after everything is done.

The issue(in the code I sent to you) is about launching the renamed file right?? It doesn't give the expected rom name so IARL matching technique cannot find a match(I know IARL is right),

I may be wrong, is recursive decompression necessary? I mean if the splitted files have the same base name with only different sequential extension/number, sorting the file list inside the directory and decompressing the first one in the list isn't all we need to merge all of them?? Of course if the custom list follows at least this rule.

About decompression in a different thread for the computer, the only feedback we have is the output the exe gives right? I don't know if showing a busy dialog or message until operation is done is safe enough, from what I saw other guys do, they use Kodi dialogs hahaha.

What do you think about Autoit3?? IMHO it can be a good alternative to batch files, it offers Dlls handle, so it is possible to make an **unrar.exe** completely stand alone, no need for modules, winrar exes, then instead of calling a batch file, you can call **unrar.exe** with parameters.

Anyway, thanks for your effort, maybe my issue has no solution cause the files are total mess with no standard. A more generalized manual post download action is a good option, the user is responsible to give a batch file or an exe to deal with his own list.

Cheers.

**Comment by NerdFlix** [comment]
Ok, I finally gave up renaming the downloaded files to a friendly name(Files downloaded from my custom list). So now I give the proper rom name IARL is expecting and the rom name is matching inside IARL matching function.

I've spent a couple of hours testing some games of my custom lists, specially splitted files, and the code I've sent to you is downloading, uncompressing, merging splitted files(if necessary), and renaming the final rom, and now the games can be launched with no problem.

You can close this issue if you want, and just in case you'd like to test the custom list I sent to you, just let me know so I PM you with the changes I've made.

You were right, **7z.exe** can extract even Rar files (even RAR5 ones) for Windows, I don't know how to solve this issue for other OS(**maybe giving the user the responsability to deal with his own list through a script or binary??**), on Windows everything is working fine. Decompression in a different thread, maybe showing a busy dialog or something to avoid the user to press any button, and at the same time letting him know that operation is still working, could be a good option.

If you're interested, the best file decompressor for Windows I found was the one made using Autoit script language(better than using batch files), it can decompress almost all known files for Windows, here is the link with source code and binaries: https://www.legroom.net/software/uniextract

Thank you for your attention and effort.
